File: LinkedLists/Doubly-LinkedList/DLinkedList.py
#Establishing Node
class Node:

    def __init__(self):
        self.data = None
        self.Previous = None
        self.Next = None

#Establishing List
class LinkedList:

    #Initializing List
    def __init__(self):
        self.head = None
        self.size = 0


    #Check whether the list is Empty or not
    def isEmpty(self):
        return self.head == None


    #Traversing The List
    def traverse(self):

        if(self.isEmpty()):
            print("List is Empty!")
            return

        currentNode = self.head
        
        while(currentNode != None):
            print(currentNode.data, end = " -> ")
            currentNode = currentNode.Next;

        print()


    #Inserting Node(s)
    def insert(self, x, loc):

        newNode = Node()
        newNode.data = x

        if loc == 0:

            newNode.Previous = self.head
            newNode.Next = self.head
            self.head = newNode;

        else:

            if loc > self.size:

                print("The list is only {} nodes".format(self.size))
                return

            currentNode = self.head

            for i in range(loc - 1):
                currentNode = currentNode.Next

            newNode.Next = currentNode.Next

            if loc < self.size - 1:
                currentNode.Next.Previous = newNode

            currentNode.Next = newNode
            newNode.Previous = currentNode

        self.size += 1


    #Deleting Node(s)
    def delete(self, loc):

        if(self.isEmpty()):

            print("List is Already Empty!")
            return

        if(loc > self.size - 1):
            print("The list is only {} nodes!".format(self.size))
            return

        currentNode = self.head

        if loc == 0:

            self.head = currentNode.Next
            currentNode.Previous = None

        elif loc == self.size - 1:
            
            for i in range(loc - 1):
                currentNode = currentNode.Next

            currentNode.Next = None
            
        else:

            for i in range(loc):

                currentNode = currentNode.Next

                
            currentNode.Previous.Next = currentNode.Next
            currentNode.Next.Previous = currentNode.Previous

        self.size -= 1

File: LinkedLists/Doubly-LinkedList/DLinkedList.java
public class DLinkedList {

    //Establishing Node
    class Node {

        int data;
        Node previous;
        Node next;

    }

    private Node head;
    private int size;

    //Initialization
    DLinkedList(){

        head = null;
        size = 0;

    }

    //Check whether the list is Empty or not
    public boolean isEmpty(){
        return head == null;
    }

    //Traversing The List
    public void traverse(){

        if(isEmpty()){
            System.out.println("List is Empty!");
            return;
        }

        Node currentNode = head;

        while(currentNode != null){
            System.out.print(currentNode.data + " -> ");
            currentNode = currentNode.next;
        }

        System.out.println();

    }

    //Inserting Node(s)
    void insert(int x, int loc){

        Node newNode = new Node();
        newNode.data = x;

        if(loc == 0){

            newNode.previous = head;
            newNode.next = head;
            head = newNode;

        } else {

            if(loc > size){
                System.out.println("The list is only" + size + " nodes!");
                return;
            }

            Node currentNode = head;

            for(int i = 0; i < loc - 1; i++)
                currentNode = currentNode.next;

            newNode.next = currentNode.next;

            if(loc < size - 1){

                currentNode.next.previous = newNode; //

            }

            currentNode.next = newNode;
            newNode.previous = currentNode;

        }

        size++;

    }

    //Delete Node(s)
    void delete(int loc){

        if(isEmpty()){

            System.out.println("List is Already Empty!");
            return;

        }

        if(loc > size - 1){

            System.out.println("The list is only" + size + " nodes");
            return;

        }

        Node currentNode = head;

        if(loc == 0){

            head = currentNode.next;
            currentNode.previous = null;

        } else if(loc == size - 1){

            for(int i = 0; i < loc - 1; i++)
                currentNode = currentNode.next;

            currentNode.next = null;

        } else {

            for(int i = 0; i < loc; i++)
                currentNode = currentNode.next;

            currentNode.previous.next = currentNode.next;
            currentNode.next.previous = currentNode.previous;

        }

        size--;

    }

}
